# **Week 11 Applied Session: Data Validation**

We will cover the main topics from week 11 seminar in this applied session: common data validation you might perform.



There are a few things that we can check to validate the data, for example:

- Structure validation
- Content validation
- Logical validation

Dataset downloaded from [here](https://www.kaggle.com/datasets/prasad22/healthcare-dataset) is used for this notebook. The dataset contains the following columns:

`Name, Age, Gender, Blood Type, Medical Condition, Date of Admission, Doctor, Hospital, Insurance Provider, Billing Amount, Room Number, Admission Type, Discharge Date, Medication, and Test Results`.


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


### Load the data

In [ ]:
import pandas as pd

In [ ]:
health_data = pd.read_csv('/content/drive/Shareddrives/FIT5196_S1_2025/week11/healthcare_dataset.csv')
health_data.shape

(55500, 15)

### 1.1 Structure Validation

The first thing we can check is the structure of the data. We can check the following:

- The number of rows and columns
- The column names
- The data types of each column
- The number of missing values in each column
- The number of unique values in each column
- The number of duplicate rows

In [ ]:
# 1. the number of rows in the dataset
print("The number of rows in the dataset is: ", health_data.shape[0])
# 2. the number of columns in the dataset
print("The number of columns in the dataset is: ", health_data.shape[1])
# 3. the data types of each column
print("The data types of each column are: ", health_data.dtypes)

The number of rows in the dataset is:  55500
The number of columns in the dataset is:  15
The data types of each column are:  Name                   object
Age                     int64
Gender                 object
Blood Type             object
Medical Condition      object
Date of Admission      object
Doctor                 object
Hospital               object
Insurance Provider     object
Billing Amount        float64
Room Number             int64
Admission Type         object
Discharge Date         object
Medication             object
Test Results           object
dtype: object


In [ ]:
# 4. the number of missing values in each column
print("The number of missing values in each column are: ", health_data.isnull().sum())

The number of missing values in each column are:  Name                  0
Age                   0
Gender                0
Blood Type            0
Medical Condition     0
Date of Admission     0
Doctor                0
Hospital              0
Insurance Provider    0
Billing Amount        0
Room Number           0
Admission Type        0
Discharge Date        0
Medication            0
Test Results          0
dtype: int64


In [ ]:
# 5. the number of unique values in each column
print("The number of unique values in each column are: ", health_data.nunique())

The number of unique values in each column are:  Name                  49992
Age                      77
Gender                    2
Blood Type                8
Medical Condition         6
Date of Admission      1827
Doctor                40341
Hospital              39876
Insurance Provider        5
Billing Amount        50000
Room Number             400
Admission Type            3
Discharge Date         1856
Medication                5
Test Results              3
dtype: int64


In [ ]:
health_data.duplicated()

,0
0,False
1,False
2,False
3,False
4,False
...,...
55495,False
55496,False
55497,False
55498,False


In [ ]:
# 6. the number of duplicate rows
print("The number of duplicate rows are: ", health_data.duplicated().sum())
# show the duplicate rows
print("The duplicate rows are: ", health_data[health_data.duplicated()])

The number of duplicate rows are:  534
The duplicate rows are:                       Name  Age  Gender Blood Type Medical Condition  \
50023        Samuel joYCe   56    Male         O+         Arthritis   
50040    KIMBErly vasqueZ   26    Male         A-           Obesity   
50055      SCoTT tHOrNton   55  Female         B-           Obesity   
50070         wiLliam LEe   27  Female         O+         Arthritis   
50078  chRistOPheR NoRrIs   38    Male        AB-         Arthritis   
...                   ...  ...     ...        ...               ...   
55461      connOR coMPTon   63    Male         A+            Asthma   
55462       alYSsA mIlLER   35  Female         A-          Diabetes   
55464        ChRIs huGHeS   35  Female        AB-           Obesity   
55484     keNNEtH alvarez   80    Male         O+            Cancer   
55491       lInDA grIFfin   47  Female         O+           Obesity   

      Date of Admission             Doctor                    Hospital  \
50023    

Here we can see that the dataset has `534` rows of duplicate data. We can remove these duplicates to avoid any issues with the analysis.

In [ ]:
# remove the duplicate rows
health_data.drop_duplicates(inplace=True)
print("The number of rows after removing the duplicate rows is: ", health_data.shape[0])

The number of rows after removing the duplicate rows is:  54966


### 1.2 Content Validation

The next thing we can check is the content of the data. We can check the following:

- The range of values for each column
- The distribution of values for each column
- The relationship between columns
- The consistency of values between columns

In [ ]:
# 1. the range of value for each column
print("The range of value for each column are: \n", health_data.describe().loc[['min', 'max']])

The range of value for each column are: 
       Age  Billing Amount  Room Number
min  13.0    -2008.492140        101.0
max  89.0    52764.276736        500.0


We can see that only the numerical columns are displayed in the above table.

In [ ]:
# 2. distribution of values for each column
print("The distribution of values for each column are: \n", health_data.describe())

The distribution of values for each column are: 
                 Age  Billing Amount   Room Number
count  54966.000000    54966.000000  54966.000000
mean      51.535185    25544.306284    301.124404
std       19.605661    14208.409711    115.223143
min       13.000000    -2008.492140    101.000000
25%       35.000000    13243.718641    202.000000
50%       52.000000    25542.749145    302.000000
75%       68.000000    37819.858159    401.000000
max       89.000000    52764.276736    500.000000


We can find that the minimum `Billing Amount` value is negative (around -2008.492140). We can check the number of rows with negative `Billing Amount` values.

In [ ]:
# check the negative billing amount
print("The number of negative billing amount is: ", health_data[health_data['Billing Amount'] < 0].shape[0])

The number of negative billing amount is:  106


In [ ]:
# show the negative billing amount
print("The negative billing amount is: \n", health_data[health_data['Billing Amount'] < 0][['Billing Amount']])

The negative billing amount is: 
        Billing Amount
132       -502.507813
799      -1018.245371
1018      -306.364925
1421      -109.097122
2103      -576.727907
...               ...
52213    -1310.272895
52629     -676.852502
53204     -306.364925
53232     -591.917419
54136     -199.663795

[106 rows x 1 columns]


<span style="color:red">How to deal with the negative billing amount values?</span>

Typically, the billing amount values are expected to be positive. However, there are instances where billing amounts can be negative. For instance, if a patient has `overpaid` their bill, the billing amount may appear as negative.

Similarly, when `refunding` money to the patient, the billing amount can be negative.

Additionally, hospitals may `adjust bills`, which can result in a negative billing amount.

<span style="color:red">Can we determine if any of the reasons mentioned above exist in this data?</span>

In [ ]:
# Lets check if there are any person with multiple records
print("The number of person with multiple records are: ", health_data['Name'].duplicated().sum())
# show the person with multiple records and negative billing amount
print("The person with multiple records and negative billing amount are: \n", health_data[(health_data['Name'].duplicated()) & (health_data['Billing Amount'] < 0)])

The number of person with multiple records are:  4974
The person with multiple records and negative billing amount are: 
                            Name  Age  Gender Blood Type Medical Condition  \
50099              tErRy WILSOn   43    Male         O-          Diabetes   
50369             LiSA rOBInSoN   33  Female         B-         Arthritis   
50694              tHoMas PrAtt   61  Female         A+           Obesity   
50890        elIzABETH thOMPson   56  Female         A-            Cancer   
52095           MAtThew siMMons   43  Female         A-            Cancer   
52213  mr. ChRIStOPhER aLvARaDO   74    Male        AB+           Obesity   
52629          gABrieLlE deCkER   66    Male        AB-         Arthritis   
53204             AsHley WaRnER   55    Male         A+      Hypertension   
53232              dAnIel dRaKe   68  Female         B+      Hypertension   
54136         DR. micHaEl MckAy   64    Male         O+            Cancer   

      Date of Admission       

In [ ]:
negative_billing_amount_person = health_data[(health_data['Name'].duplicated()) & (health_data['Billing Amount'] < 0)]['Name'].unique()
print("The person with multiple records and negative billing amount are: \n", negative_billing_amount_person)

The person with multiple records and negative billing amount are: 
 ['tErRy WILSOn' 'LiSA rOBInSoN' 'tHoMas PrAtt' 'elIzABETH thOMPson'
 'MAtThew siMMons' 'mr. ChRIStOPhER aLvARaDO' 'gABrieLlE deCkER'
 'AsHley WaRnER' 'dAnIel dRaKe' 'DR. micHaEl MckAy']


In [ ]:
# get one example
health_data[(health_data['Name'] == negative_billing_amount_person[0])]

,Name,Age,Gender,Blood Type,Medical Condition,Date of Admission,Doctor,Hospital,Insurance Provider,Billing Amount,Room Number,Admission Type,Discharge Date,Medication,Test Results
29484,tErRy WILSOn,39,Male,O-,Diabetes,2024-02-19,Diana Smith,"Medina and Elliott Stewart,",Aetna,-1316.618581,491,Emergency,2024-03-20,Paracetamol,Inconclusive
50099,tErRy WILSOn,43,Male,O-,Diabetes,2024-02-19,Diana Smith,"Medina and Elliott Stewart,",Aetna,-1316.618581,491,Emergency,2024-03-20,Paracetamol,Inconclusive


In [ ]:
health_data[(health_data['Name'] == negative_billing_amount_person[1])]

,Name,Age,Gender,Blood Type,Medical Condition,Date of Admission,Doctor,Hospital,Insurance Provider,Billing Amount,Room Number,Admission Type,Discharge Date,Medication,Test Results
14534,LiSA rOBInSoN,29,Female,B-,Arthritis,2022-12-29,Nancy Thompson,Chandler PLC,Cigna,-26.112523,198,Emergency,2023-01-12,Penicillin,Normal
50369,LiSA rOBInSoN,33,Female,B-,Arthritis,2022-12-29,Nancy Thompson,Chandler PLC,Cigna,-26.112523,198,Emergency,2023-01-12,Penicillin,Normal


__It seems that the 'Age' also has a mismatch issue regarding the above records__

### 1.3 Logical Validation

The last thing we can check is the logic of the data. This check involves checking the data against business rules or domain knowledge.

In [ ]:
# check if there is negative age
print("The number of negative age is: ", health_data[health_data['Age'] < 0].shape[0])

The number of negative age is:  0


In [ ]:
# check the Blood Type column, and show the unique values to see if there is any value that is not in the specified values
print("The unique values in Blood Type column are: ", health_data['Blood Type'].unique())

The unique values in Blood Type column are:  ['B-' 'A+' 'A-' 'O+' 'AB+' 'AB-' 'B+' 'O-']


In [ ]:
# check if there is any row Date of Admission is later than Date of Discharge
print("The number of rows where Date of Admission is later than Date of Discharge is: \n", health_data[health_data['Date of Admission'] > health_data['Discharge Date']].shape[0])

The number of rows where Date of Admission is later than Date of Discharge is: 
 0
